### PR TITLE
Ensure that users only receive notifications for their courses

### PR DESCRIPTION
**Pull request by sascha11110**
### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.

### Motivation and Context
Some users receive recent notifications for courses to which they do not belong.

### Description
The REST route `/notifications` related to `findAllExceptSystem()` in the `NotficationService` seems to work as the *unwanted* notifications are not displayed there, only in the `notification-container.component`. 

Therefore I adapted the implementation of `findAllRecentExceptSystem()` in the `NotficationService` that is related to the REST route `/notifications/for-user` as closely as possible to the implementation of `findAllExceptSystem()`.

I also added several test cases for the above mentioned rest routes.

Additionally, I made sure that the notification service gets cleaned up when the user logs out.

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Administration
3. Click on the notification bell in the top navigation
4. Check that the displayed notifications only belong to courses of the current user

### CODE REVIEW

File: src/main/webapp/app/overview/notification/notification.service.ts [hunk 1]
```diff
@@ -183,4 +174,13 @@ export class NotificationService {
         const courseId = target.course || notification.course.id;
         this.router.navigate([target.mainPage, courseId, target.entity, target.id]);
     }
+
+    private initNotificationObserver(): void {
+        this.notificationObserver = new BehaviorSubject<Notification | null>(null);
+    }
+
+    public cleanUp(): void {
+        this.cachedNotifications = new Observable<EntityArrayResponseType>();
```
Comment (krusche) [inline src/main/webapp/app/overview/notification/notification.service.ts:183]: I guess we should cleanup `subscribedTopics` as well, right?

Comment (sascha11110) [reply src/main/webapp/app/overview/notification/notification.service.ts:183]: Had that on my todo :-) Is now cleaned up during logout: https://github.com/ls1intum/Artemis/pull/1323/commits/0af1890947ced7227abc9f4e7db24fe3ed1a5505

**Comment by sascha11110** [comment]
I've added several test cases to `NotificationResourceIntegrationTest` and `subscribedTopics` will also be cleaned up when the user logs out.

**Comment by madwau** [comment]
Code-wise everything looks good to me. I tried testing it on the test server but wasn't really able to create scenarios that would properly confirm if the change does what it is supposed to do.

File: src/test/java/de/tum/in/www1/artemis/NotificationResourceIntegrationTest.java [hunk 2]
```diff
@@ -101,36 +115,132 @@ public void testCreateNotification_asInstructor_BAD_REQUEST() throws Exception {
...
+        SingleUserNotification notification2 = ModelFactory.generateSingleUserNotification(ZonedDateTime.now(), users.get(1));
+        notificationRepository.save(notification2);
+
+        List<Notification> notifications = request.getList("/api/notifications", HttpStatus.OK, Notification.class);
+        assertThat(notifications).as("Notification with recipient equal to current user is returned").contains(notification1);
+        assertThat(notifications).as("Notification with recipient not equal to current user is not returned").doesNotContain(notification2);
+
+        List<Notification> recentNotifications = request.getList("/api/notifications/recent-for-user", HttpStatus.OK, Notification.class);
+        assertThat(recentNotifications).as("Recent notification with recipient equal to current user is returned").contains(notification1);
+        assertThat(recentNotifications).as("Recent notification with recipient not equal to current user is not returned").doesNotContain(notification2);
```
[the remark this answers is not in the document]
Comment (sascha11110) [reply src/test/java/de/tum/in/www1/artemis/NotificationResourceIntegrationTest.java:133]: As discussed on zoom this is tested implicitly via `testGetRecentNotifications()`.

**Comment by sascha11110** [comment]
I fixed the issue found by @sleiss in https://github.com/ls1intum/Artemis/pull/1323/commits/b32616a45df11759e49d39355648cb7f401ce572.

Thanks to @sleiss and @F4lka for your support!

**Pull request merged**